src/core/opamHash.ml

```diff
@@ -0,0 +1,97 @@
+(**************************************************************************)
+(*                                                                        *)
+(*    Copyright 2016 OCamlPro                                             *)
+(*                                                                        *)
+(*  All rights reserved. This file is distributed under the terms of the  *)
+(*  GNU Lesser General Public License version 2.1, with the special       *)
+(*  exception on linking described in the file LICENSE.                   *)
+(*                                                                        *)
+(**************************************************************************)
+
+type kind = [ `MD5 | `SHA1 | `SHA256 ]
+
+let default_kind = `MD5
+
+type t = kind * string
+
+let kind = fst
+
+let pfx_sep_char = '='
+let pfx_sep_str = String.make 1 pfx_sep_char
+
+let string_of_kind = function
+  | `MD5 -> "md5"
+  | `SHA1 -> "sha1"
+  | `SHA256 -> "sha256"
+
+let kind_of_string s = match String.lowercase s with
+  | "md5" -> `MD5
+  | "sha1" -> `SHA1
+  | "sha256" -> `SHA256
+  | _ -> invalid_arg "OpamHash.kind_of_string"
+
+let is_hex_str len =
+  Re.execp Re.(compile (repn xdigit len (Some len)))
+
+let len = function
+  | `MD5 -> 32
+  | `SHA1 -> 40
+  | `SHA256 -> 64
+
+let valid kind = is_hex_str (len kind)
+
+let make kind s =
+  if valid kind s then kind, String.lowercase s
+  else invalid_arg ("OpamHash.make_"^string_of_kind kind)
+
+let md5 = make `MD5
+let sha1 = make `SHA1
+let sha256 = make `SHA256
+
+let of_string_opt s =
+  try
+    let kind, s =
+      match OpamStd.String.cut_at s pfx_sep_char with
+      | None -> `MD5, s
+      | Some (skind, s) -> kind_of_string skind, s
+    in
+    if valid kind s then Some (kind, String.lowercase s)
+    else None
+  with Invalid_argument _ -> None
+
+let of_string s =
+  match of_string_opt s with
+  | Some h -> h
+  | None -> invalid_arg "OpamHash.of_string"
+
+let to_string (kind,s) =
+  String.concat pfx_sep_str [string_of_kind kind; s]
+
+let to_json s = `String (to_string s)
+
+let to_path (kind,s) =
+  [string_of_kind kind; String.sub s 0 2; s]
+
+let compute ?(kind=default_kind) file = match kind with
+  | `MD5 -> md5 (Digest.to_hex (Digest.file file))
+  | `SHA1 | `SHA256 ->
+    match
+      OpamSystem.read_command_output ["openssl"; string_of_kind kind; file]
```

Comment:
but as mentioned earlier, this can be solved by a `download-command` like configuration option which by default uses `openssl md5 %file% | cut -d '=' -f 2` ;)